### DISCUSSION TITLE
How does Materials object call member functions in VectorPostprocessor object?

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#21369] [Q&A General] [answered]
Dear MOOSE experts,

Recently I wanted to construct a grain boundary energy as a function of grain size in the material object `GBEvolution`. By searching the code, I found only the case where the material class is coupled with the userobject, like: `_grain_tracker(getUserObject<GrainTracker>("grain_tracker")),` but not the case where the VectorPostprocessor is called. In addition, the `FeatureVolumeVectorPostprocessor` object calculates the volume of each feature, and creates the pubilc function `getFeatureVolume` so that we can get the volume of each feature based on the feature id.

So my question is, how to call `FeatureVolumeVectorPostprocessor.getFeatureVolume` in material object `GBEvolution` to get feature volume.

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by GiudGiud** [comment, marked as answer]
Hello

If this is a standard vectorpostprocessor you should ` _fe_problem.getVectorPostprocessorValueByName(_vpp_name, vec_name);`
to retrieve the values of the VPP. You can bind this to a reference in the constructor so that you only call this once.

Guillaume

**Reply by PengWei97** [reply]
Ok, Thank you for your very kind reply.

This function `getVectorPostprocessorValueByName` can achieve my requirements, with the following code,
```cpp
  const auto & k1 = _fe_problem.getVectorPostprocessorValueByName(_vector_postprocessor_name, _vector_name); // _fe_problem. VectorPostprocessorValue 
  std::cout << "the size of k1 is " << k1.size() <<std::endl;
  for(unsigned int i = 0; i < k1.size(); i++)
      std::cout << "the value of k1 is " << k1[i] <<std::endl;  
```
But I'm not quite sure why `_fe_problem` is added? Please can you give me guidance on this.
Similarly, if you do not use `_fe_problem`, you can also get the desired result, and the result is as follows.
![image](https://user-images.githubusercontent.com/56183775/175042841-6e4c75a6-451f-4709-84b9-92daebbfc23c.png)

wei

**Reply by GiudGiud** [reply]
it depends on which object has the VPPInterface to use this routine. The problem object does. I guess your material also does.
Either way is fine

**Reply by PengWei97** [reply]
good, the problem is solved, thank you very much!
